### DISCUSSION TITLE
why i can't connect iotdb 0.12.1 service by pyspark?

### DISCUSSION THREAD

**Discussion by tvboy20082018** [apache/iotdb#3851] [Q&A] [answered]
i have tried to connect iotdb 0.12.1 service by pyspark with all kinds of methods, also i'm sure i have written python code correctly and executed spark-submit commands with params correctly. however, i have been receiving the spark-submit execution result with "cannot establish connection with jdbc:iotdb://172.24.4.42:6668/" and the server error with "org.apache.thrift.transport.TTransportException: Read a negative frame size" all the time.  I've been tortured by this problem for a whole day. Could somebody give a hand please?
![21f07550f6042c2dbc07bd95e7301b4](https://user-images.githubusercontent.com/40767885/131065652-998a7963-67f8-4099-8b2a-a0b07c8f7812.png)
![87d5fbb0b9194dd6c98c373c12a2f23](https://user-images.githubusercontent.com/40767885/131065742-123ebc92-1645-490e-95de-e1c06fd7722b.png)
![64bb7915ce33a0e5cfa0cc48b558cb8](https://user-images.githubusercontent.com/40767885/131065675-d1610f27-a37b-4c8a-9254-331ba656396f.png)
![8b83a55ce2e506a14c86317e2c5c77c](https://user-images.githubusercontent.com/40767885/131065695-24d6eee0-6880-4b44-8d36-d06049ba67f4.png)

**Comment by SilverNarcissus** [comment, marked as answer]
From you error log. The cause of this error is the port of JDBC connection. If you use default config, you should connect to port NO.6667 rather than 6668. 
For trouble shooting, you can firstly use command line or java example to connect iotdb without spark. Then you can try use spark locally in our spark example. Finally, submit this job by spark-submit. During these step you can figure out where is the problem.

**Reply by tvboy20082018** [reply]
Sorry, i attached wrong shot with python code, in fact the right jdbcurl = "jdbc:iotdb://172.24.4.42:6668/".
i've configured rpc_address and rpc_port for iotdb 0.12.1 service. it has been okay that i connected to iotdb 0.12.1 service by iotdb 0.12.1 client.
i've also followed your guide to use jdbc to connect iotdb 0.12.1 service without spark successfully, but still couldn't connect it with pyspark.
do you think maybe there is something wrong with the protocol between thrift and spark-iotdb-connector?
![image](https://user-images.githubusercontent.com/40767885/131082625-ae36412b-9867-407b-9e2a-81c970169ec0.png)
![image](https://user-images.githubusercontent.com/40767885/131082671-a5b8422b-c7fb-439a-ae01-2312005724ce.png)
![image](https://user-images.githubusercontent.com/40767885/131082963-9f5783bc-5297-4ce4-ae3e-7a23685d54b7.png)
![image](https://user-images.githubusercontent.com/40767885/131084338-967e980e-622a-42a4-885a-7137f92b6a04.png)

**Comment by SilverNarcissus** [comment]
I will try it later. You can try with spark-shell as follow:
```
spark-shell --jars [your jars path]

import org.apache.iotdb.spark.db._

val df = spark.read.format("org.apache.iotdb.spark.db").option("url","jdbc:iotdb://127.0.0.1:6667/").option("sql","select * from root").load

df.printSchema()

df.show()
```

**Reply by tvboy20082018** [reply]
i followed official case for spark iotdb  with scala
![de53ea2722dbca812bd69e04c40dd8b](https://user-images.githubusercontent.com/40767885/131094340-57030d0c-f55b-406e-9a9f-51893c5f9b0a.png)
however, i got the same result with pyspark-connecting-iotdb: java.sql.SQLException: Can not establish connection with jdbc:iotdb://XXX
![f37953a7fa6d68a17fb900c1d534839](https://user-images.githubusercontent.com/40767885/131094454-20f133e9-a675-45c2-92c7-4a25db0dc7c0.png)
i'm confused where i was stuck.

**Reply by SilverNarcissus** [reply]
I have tried it and it can work on my computer. You may use 0.11 iotdb to try again?

**Reply by tvboy20082018** [reply]
my comrade replaced jars of the right version on Spark, then it worked and the error "java.sql.SQLException: Can not establish connection with jdbc:iotdb://172.24.4.42:6668/" didn't occur. however, another problem happened that "WARN TaskSetManager: Lost task 0.0 in stage 0.0 (TID 0, node03, executor 2): org.apache.spark.util.TaskCompletionListenerException: null" occurs in the output of spark-submit execution. i'm confused with the problem and have no clue about it.

**Reply by tvboy20082018** [reply]
![1630117165(1)](https://user-images.githubusercontent.com/40767885/131203283-bafca953-67f8-46e2-aaef-bf52cb15aba4.jpg)
![1630117492(1)](https://user-images.githubusercontent.com/40767885/131203312-3e3f1327-2f4c-4fef-ba32-3c0d30cb9a1a.png)

**Reply by SilverNarcissus** [reply]
You can also run it in local mode without submit. Is there any problem on the iotdb server?

**Comment by tvboy20082018** [comment]
I've solved the problem by replacing spark cluster server and using spark-iotdb-connector-0.12.0 to connect iotdb 0.12.1. thanks for SilverNarcissus and qiaojialin's help and patience!
